source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,659 @@
+  if (isScope!Scope)
+{
+  return refract!(
+    __traits(getOverloads, Scope, name)[index],
+    `__traits(getOverloads, %s, "%s")[%d]`.format(localSymbol, name, index))
+    .setIndex(index);
+}
+
+private enum True(T...) = true;
+
+Function[] refract(alias Scope, string localSymbol, alias Predicate = True)()
+  if (isScope!Scope)
+{
+  Function[] functions;
+
+  static foreach (member; __traits(allMembers, Scope)) {
+    static foreach (index, fun; __traits(getOverloads, Scope, member)) {
+      static if (Predicate!fun) {
+        functions ~= refract!(Scope, localSymbol, member, index);
+      }
+    }
+  }
+
+  return functions;
+}
+
+///
+unittest
+{
+  pure @nogc int answer(lazy string question);
+  alias F = answer; // typically F is a template argument
+  static assert(
+    refract!(F, "F").mixture ==
+    "pure @nogc @system std.traits.ReturnType!(F) answer(lazy std.traits.Parameters!(F)[0] _0);");
+}
+
+///
+unittest
+{
+  import std.format;
+  import std.traits;
+
+  interface GrandTour
+  {
+    pure int foo() immutable;
+    @nogc @trusted nothrow ref int foo(out real, return ref int, lazy int) const;
+    @safe shared scope void bar(scope Object);
+  }
+
+  class Mock(Interface) : Interface
+  {
+    static foreach (member; __traits(allMembers, Interface)) {
+        static foreach (fun; __traits(getOverloads, Interface, member)) {
+          mixin({
+              enum Model = refract!(fun, "fun");
+              if (is(ReturnType!fun == void)) {
+                return Model.setBody("{}").mixture;
+              } else if (Model.attributes & FunctionAttribute.ref_) {
+                return Model.setBody(q{{
+                    static %s rv;
+                    return rv;
+                  }}.format(Model.returnType)).mixture;
+              } else {
+                return Model.setBody(q{{
+                    return %s.init;
+                  }}.format(Model.returnType)).mixture;
+              }
+            }());
+        }
+    }
+  }
+
+  GrandTour mock = new Mock!GrandTour;
+  real x;
+  int i, l;
+  mock.foo(x, i, l++) = 1;
+  assert(mock.foo(x, i, l++) == 1);
+  assert(l == 0);
+}
+
+private enum isAggregate(T...) =
+  is(T[0] == struct) || is(T[0] == union) || is(T[0] == class)
+  || is(T[0] == interface);
+
+/**
+    A struct capturing all the properties of a function.
+*/
+
+private mixin template replaceAttribute(string Name)
+{
+  alias Struct = typeof(this);
+  mixin(
+    "Struct copy = {",
+    {
+      string[] mixture;
+      foreach (member; __traits(allMembers, Struct)) {
+        if (__traits(getOverloads, Struct, member).length == 0) {
+          //pragma(msg, member);
+          mixture ~= member ~ ":" ~ (member == Name ? "value" : member);
+        }
+      }
+      return mixture.join(",\n");
+    }(),
+    "};"
+  );
+}
+
+
+immutable struct Function
+{
+
+  /**
+     A string that evaluates to a function symbol.
+   */
+
+  string localSymbol;
```

Comment:
Should this be accessible publicly? This is the string that's provided during refraction right?

Reply:
> Should this be accessible publicly? This is the string that's provided during refraction right?

Hmmm it is a *private* helper that is tested many times implicitly but why not...

Reply:
Why not make it public? I don't know. If there's no immediate use for making it public I wouldn't.